### PR TITLE
support volume partition

### LINKED ISSUE #812: Volume partitions are not supported
/kind bug

**What happened?**
For CSI migration we need all in-tree features supported in the CSI driver. In-tree volumes support mounting volume partitions in  `pv.spec.awsElasticBlockStore.partition` flag. Such capability should be present in the CSI driver too.

**How to reproduce it (as minimally and precisely as possible)?**
Enable CSI migration and use following Pod:

```
apiVersion: v1
kind: Pod
metadata:
  name: inline
spec:
  containers:
  - name: inline
    image: centos:7
    args:
    - sleep
    - '99999'
    volumeMounts:
    - mountPath: "/mnt/storage"
      name: inline
  volumes:
  - name: inline
    awsElasticBlockStore:
      fsType: xfs
      volumeID: vol-072b2b73e088c43b5
      partition: 1
```

**Anything else we need to know?**:
The CSI translation library makes the partition number available in `volume_attribute: partition`: https://github.com/kubernetes/kubernetes/blob/bbb58fa085ccfc4635aa8544686d960115449d95/staging/src/k8s.io/csi-translation-lib/plugins/aws_ebs.go#L112

### PR DESCRIPTION
**Pull request by AndyXiangLi**
**Is this a bug fix or adding new feature?**
Fixes #812 
**What is this PR about? / Why do we need it?**
Add volume partition support when stage&publish volumes to sync with in-tree driver
**What testing is done?** 
unit test
e2e test

### CODE REVIEW

File: pkg/driver/node.go [hunk 1]
```diff
@@ -521,7 +525,13 @@ func (d *nodeService) nodePublishVolumeForBlock(req *csi.NodePublishVolumeReques
 	if !exists {
 		return status.Error(codes.InvalidArgument, "Device path not provided")
 	}
-	source, err := d.findDevicePath(devicePath, volumeID)
+
+	partition := ""
```
Comment (wongma7) [inline pkg/driver/node.go:529]: partition would not make sense in block context, like block that has been partitioned is not block. we should throw an error imo if we find the attribute

File: pkg/driver/node_linux.go [hunk 2]
```diff
@@ -49,7 +52,15 @@ func (d *nodeService) findDevicePath(devicePath, volumeID string) (string, error
 	// /dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0fab1d5e3f72a5e23
 	nvmeName := "nvme-Amazon_Elastic_Block_Store_" + strings.Replace(volumeID, "-", "", -1)
 
-	return findNvmeVolume(nvmeName)
+	nvmeDevicePath, err := findNvmeVolume(nvmeName)
+	if err != nil {
+		//Return the nvmeDevicePath just for logging purpose
```
Comment (wongma7) [inline pkg/driver/node_linux.go:57]: the callers never read the returned string though.

File: pkg/driver/node.go (hunk 1 above)
Comment (wongma7) [reply pkg/driver/node.go:529]: nvm, i think you can still partition a device wthout formatting any of the partitions right ? Not an expert on this obviously

**Comment by wongma7** [comment]
Can we add an e2e test otherwise lgtm

File: pkg/driver/node.go (hunk 1 above)
Comment (wongma7) [reply pkg/driver/node.go:529]: just checked in-tree and the code is same except in-tree never bothered to set partition of nvme volumes : / with "p#" suffix

Comment (AndyXiangLi) [reply pkg/driver/node.go:529]: in-tree is going to have nvme fix in this PR https://github.com/kubernetes/kubernetes/pull/100500
I'm testing the block scenario and will update here later.

File: pkg/driver/node_linux.go (hunk 2 above)
Comment (AndyXiangLi) [reply pkg/driver/node_linux.go:57]: u r right... It is logging devicePath, will return empty string then

File: pkg/driver/node.go (hunk 1 above)
Comment (AndyXiangLi) [reply pkg/driver/node.go:529]: @wongma7 just to confirm we can still partition a device without formatting.

**Comment by AndyXiangLi** [comment]
@wongma7 @jsafrane Update the PR as migration test was failing. Notice that some migration test cases set partition as "0", so explicitly check that scenario and ignore partition 0. Please LMK if you have any idea on this

File: pkg/driver/node.go [hunk 3]
```diff
@@ -521,7 +529,17 @@ func (d *nodeService) nodePublishVolumeForBlock(req *csi.NodePublishVolumeReques
...
+	if part, ok := req.GetVolumeContext()[VolumeAttributePartition]; ok {
+		if part != "0" {
+			partition = part
```
Comment (wongma7) [inline pkg/driver/node.go:536]: should we validate it? Check that it's an integer? and not something crazy/malicious with weird paths and whatnot.

Comment (AndyXiangLi) [reply pkg/driver/node.go:536]: Yeah, parse to int makes sense to me, is there any special rule for windows partition validation? should that be same?

Comment (wongma7) [reply pkg/driver/node.go:536]: You can leave it same, I'll test it out later.

I think I will make the windows findDevicePath implementation ignore the partition arg. From my understanding, the windows mount code will always mount the first partition so it doesn't matter what partition is set to. https://github.com/kubernetes-csi/csi-proxy/blob/d47b9ecde8f020dd9d5eb2df3e56bf2f10a3f4a3/internal/os/volume/api.go#L87 (similarly, fstype gets ignored in windows world)

**Comment by wongma7** [comment]
/lgtm
thanks

**Pull request merged**